packages/twilio-run/src/config/start.ts

```diff
@@ -67,12 +69,73 @@ export async function getUrl(cli: StartCliFlags, port: string | number) {
   let url = `http://localhost:${port}`;
   if (typeof cli.ngrok !== 'undefined') {
     debug('Starting ngrok tunnel');
-    const ngrokConfig: NgrokConfig = { addr: port };
+    // Setup default ngrok config, setting the protocol and the port number to
+    // forward to.
+    const defaultConfig: INgrokOptions = { addr: port, proto: 'http' };
+    let tunnelConfig = defaultConfig;
+    let ngrokConfig;
+    if (typeof cli.ngrokConfig === 'string') {
+      // If we set a config path then try to load that config. If the config
+      // fails to load then we'll try to load the default config instead.
+      const configPath = join(cli.cwd || process.cwd(), cli.ngrokConfig);
+      try {
+        ngrokConfig = parse(await readFilePromise(configPath, 'utf-8'));
+      } catch (err) {
+        logger.warn(`Could not find ngrok config file at ${configPath}`);
+      }
+    }
+    if (!ngrokConfig) {
+      // Try to load default config. If there is no default config file, set
+      // `ngrokConfig` to be an empty object.
+      const configPath = join(homedir(), '.ngrok2', 'ngrok.yml');
+      try {
+        ngrokConfig = parse(await readFilePromise(configPath, 'utf-8'));
+      } catch (err) {
+        ngrokConfig = {};
+      }
+    }
+    if (
+      typeof cli.ngrokName === 'string' &&
+      typeof ngrokConfig.tunnels === 'object'
+    ) {
+      // If we've asked for a named ngrok tunnel and there are available tunnels
+      // in the config, then set the `tunnelConfig` to the options from the
+      // config, overriding the addr and proto to the defaults.
+      tunnelConfig = { ...ngrokConfig.tunnels[cli.ngrokName], ...tunnelConfig };
+      if (!tunnelConfig) {
+        // If the config does not include the named tunnel, then set it back to
+        // the default options.
+        logger.warn(
+          `Could not find config for named tunnel "${cli.ngrokName}". Falling back to other options.`
+        );
+        tunnelConfig = defaultConfig;
+      }
+    }
+    if (typeof ngrokConfig.authtoken === 'string') {
+      // If there is an authtoken in the config, add it to the tunnel config.
+      tunnelConfig.authToken = ngrokConfig.authtoken;
+    }
     if (typeof cli.ngrok === 'string' && cli.ngrok.length > 0) {
-      ngrokConfig.subdomain = cli.ngrok;
+      // If we've asked for a custom subdomain, override the tunnel config with
+      // it.
+      tunnelConfig.subdomain = cli.ngrok;
+    }
+    const ngrok = require('ngrok');
+    try {
+      // Try to open the ngrok tunnel.
+      url = await ngrok.connect(tunnelConfig);
```

Comment:
Should we not just use `configPath` option passed into ngrok.connect and avoid trying to manually parse out params for ngrok? We could definitely do a fs.exist(...) on the file to ensure presence, but we should not be required to fallback and load the default ngrok file stored in the home location (that's free by default).

Reply:
We want the ngrok tunnel to connect to the http server we're starting on the port we're starting it on. The config options could be trying to setup a tcp tunnel to a different port. So, I chose to parse the config and turn it into something that will work within the context of a `twilio-run` powered application.

Note, even if you create a named tunnel that points to port 3000, if we start `twilio-run` and something else is already running on port 3000 we offer a new random port number to start on instead. Parsing and manipulating the config in this way will result in a correct tunnel in that case, which would be the expected behaviour.

My idea here is that the tunnel we create will always point to the server we are running, even if it overrides the actual config, because that is what the user would expect to happen.